### PR TITLE
Workflow REST API and Replay/Retry/Error handling

### PR DESCRIPTION
**Pull request by ahgittin**
Completes almost all of what we wanted for V1 release of this feature, including good UI support (PR to follow, will reference this) for showing workflow steps and retrying (replaying) manually, and extensive support for retrying automatically, configuring retries, and configuring error handling.

### CODE REVIEW

File: rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/EntityResource.java [hunk 1]
```diff
@@ -358,4 +362,42 @@ public List<Object>  getSpecList(String applicationId, String entityId) {
         List<SpecSummary> specTag = BrooklynTags.findSpecHierarchyTag(entity.tags().getTags());
         return (List<Object>) resolving(specTag).preferJson(true).resolve();
     }
+
+    @Override
+    public List<WorkflowExecutionContext> getWorkflows(String applicationId, String entityId) {
```
Comment (jcabrerizo) [inline rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/EntityResource.java:367]: I think this new endpoint as the same as the other should check for entitlements
```
// getWorkflows and replayWorkflow
        if (Entitlements.isEntitled(mgmt().getEntitlementManager(), Entitlements.SEE_ENTITY, entity)) {
            return EntityTransformer.entitySummary(entity, ui.getBaseUriBuilder());
        }
// replayWorkflow
        if (Entitlements.isEntitled(mgmt().getEntitlementManager(), Entitlements.REPLAY_WORKFLOW, entity)) {
            return EntityTransformer.entitySummary(entity, ui.getBaseUriBuilder());
        }

```
replayWorkflow could be also authorised by `Entitlements.INVOKE_EFFECTOR` instead creating a new entitlement

File: core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowErrorHandling.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        WorkflowErrorHandlingResult result = new WorkflowErrorHandlingResult();
+        Task handlerParent = Tasks.current();
+        log.debug("Starting "+handlerParent.getDisplayName()+" with "+ errorOptions.size()+" handler"+(errorOptions.size()!=1 ? " options" : "")+" in task "+handlerParent.getId());
+
+        for (int i = 0; i< errorOptions.size(); i++) {
+            // go through steps, find first that matches
+
+            WorkflowStepDefinition errorStep = errorOptions.get(i);
+
+            WorkflowStepInstanceExecutionContext handlerContext = new WorkflowStepInstanceExecutionContext(stepIndexIfStepErrorHandler!=null ? stepIndexIfStepErrorHandler : -3, errorStep, context);
```
[the remark this answers is not in the document]
Comment (jcabrerizo) [reply core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowErrorHandling.java:100]: I found later on my review the response in WorkflowExecutionContext.java

```
    private String indexCode(int index) {
        // these numbers shouldn't be used for much, but they are used in a few places :(
        if (index==-1) return "start";
        if (index==-2) return "end";
        if (index==-3) return "error-handler";
        return "neg-"+(index); // unknown
    }
```

Sting thinking creating named constants for those special steps index would be better

[the remark this answers is not in the document]
Comment (ahgittin) [reply core/src/main/java/org/apache/brooklyn/core/workflow/WorkflowErrorHandling.java:100]: 👍

**Comment by ahgittin** [comment]
addressed everything but constants; merging

**Comment by ahgittin** [comment]
entitlements (and expiry) will be done in a following PR

**Pull request merged**